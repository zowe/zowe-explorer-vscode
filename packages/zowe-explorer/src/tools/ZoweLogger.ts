/**
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

/* eslint-disable @typescript-eslint/restrict-plus-operands */

import * as vscode from "vscode";
import * as loggerConfig from "../../log4jsconfig.json";
import * as path from "path";
import { Gui, imperative, MessageSeverity, ZoweVsCodeExtension } from "@zowe/zowe-explorer-api";

export class ZoweLogger {
    public static zeOutputChannel: vscode.OutputChannel;
    public static log: imperative.Logger;
    private static defaultLogLevel = "INFO";
    private static zeLogLevel: string;

    private static impLogger: imperative.Logger;

    public static async initializeZoweLogger(context: vscode.ExtensionContext): Promise<string> {
        try {
            const logsPath: string = ZoweVsCodeExtension.customLoggingPath ?? context.extensionPath;
            this.initializeImperativeLogger(logsPath);
            return logsPath;
        } catch (err) {
            // Don't log error if logger failed to initialize
            if (err instanceof Error) {
                const errorMessage = vscode.l10n.t("Error encountered while activating and initializing logger");
                await Gui.errorMessage(`${errorMessage}: ${err.message}`);
            }
        }
    }

    /**
     * Initializes Imperative Logger
     * @param logsPath File path for logs folder defined in preferences
     */
    private static initializeImperativeLogger(logsPath: string): void {
        const zeLogLevel = ZoweLogger.getLogSetting();
        const loggerConfigCopy = JSON.parse(JSON.stringify(loggerConfig));
        for (const appenderName of Object.keys(loggerConfigCopy.log4jsConfig.appenders)) {
            loggerConfigCopy.log4jsConfig.appenders[appenderName].filename = path.join(
                logsPath,
                loggerConfigCopy.log4jsConfig.appenders[appenderName].filename
            );
            loggerConfigCopy.log4jsConfig.categories[appenderName].level = zeLogLevel;
        }
        imperative.Logger.initLogger(loggerConfigCopy);
        this.impLogger = imperative.Logger.getAppLogger();
    }

    public static trace(message: string): void {
        this.writeLogMessage(message, MessageSeverity.TRACE);
    }

    public static debug(message: string): void {
        this.writeLogMessage(message, MessageSeverity.DEBUG);
    }

    public static info(message: string): void {
        this.writeLogMessage(message, MessageSeverity.INFO);
    }

    public static warn(message: string): void {
        this.writeLogMessage(message, MessageSeverity.WARN);
    }

    public static error(message: string): void {
        this.writeLogMessage(message, MessageSeverity.ERROR);
    }

    public static fatal(message: string): void {
        this.writeLogMessage(message, MessageSeverity.FATAL);
    }

    public static disposeZoweLogger(): void {
        this.zeOutputChannel.dispose();
    }

    public static getLogSetting(): string {
        this.zeLogLevel = vscode.workspace.getConfiguration("zowe").get("logger");
        return this.zeLogLevel ?? this.defaultLogLevel;
    }

    public static get imperativeLogger(): imperative.Logger {
        return this.impLogger;
    }

    private static writeLogMessage(message: string, severity: MessageSeverity): void {
        // Do not use Imperative logger if it hasn't been initialized.
        if (ZoweLogger.imperativeLogger == null) {
            return;
        }

        if (+MessageSeverity[this.getLogSetting()] <= +severity) {
            const severityName = MessageSeverity[severity];
            this.imperativeLogger[severityName?.toLowerCase()](message);
            this.zeOutputChannel?.appendLine(this.createMessage(imperative.Censor.censorRawData(message), severityName));
        }
    }

    private static createMessage(msg: string, level: string): string {
        return `[${this.getDate()} ${this.getTime()}] [${level}] ${msg}`;
    }

    private static getDate(): string {
        const dateObj = new Date(Date.now());
        const day = ("0" + dateObj?.getDate()).slice(-2);
        const month = ("0" + (dateObj?.getMonth() + 1)).slice(-2);
        return `${dateObj.getFullYear()}/${month}/${day}`;
    }

    private static getTime(): string {
        const dateObj = new Date(Date.now());
        const hours = dateObj?.getHours().toString().padStart(2, "0");
        const minutes = dateObj?.getMinutes().toString().padStart(2, "0");
        const seconds = dateObj?.getSeconds().toString().padStart(2, "0");
        return `${hours}:${minutes}:${seconds}`;
    }
}
