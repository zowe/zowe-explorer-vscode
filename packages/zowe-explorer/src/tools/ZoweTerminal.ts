/**
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { imperative } from "@zowe/zowe-explorer-api";
import * as vscode from "vscode";

export class ZoweTerminal implements vscode.Pseudoterminal {
    public static readonly mTermX = ">";
    public static readonly invalidChar = "ï¿½";
    public static readonly Keys = {
        EMPTY_LINE: `${this.mTermX} `,
        CLEAR_ALL: "\x1b[2J\x1b[3J\x1b[;H",
        CLEAR_LINE: `\x1b[2K\r`,
        END: "\x1b[F",
        HOME: "\x1b[H",
        CMD_LEFT: "\x01", // MacOS HOME
        CTRL_C: "\x03",
        CTRL_D: "\x04",
        CMD_RIGHT: "\x05", // MacOS END
        CTRL_BACKSPACE: "\x08",
        TAB: "\x09",
        CMD_BACKSPACE: "\x15",
        OPT_BACKSPACE: "\x17",
        OPT_CMD_BACKSPACE: "\x1b\x7F",
        BACKSPACE: "\x7f",
        INSERT: "\x1b[2~",
        DEL: "\x1b[3~",
        PAGE_UP: "\x1b[5~",
        PAGE_DOWN: "\x1b[6~",
        ENTER: "\r",
        NEW_LINE: "\r\n",
        OPT_LEFT: "\x1bb",
        OPT_RIGHT: "\x1bf",
        UP: "\x1b[A",
        DOWN: "\x1b[B",
        RIGHT: "\x1b[C",
        LEFT: "\x1b[D",
        hasModKey: (key: string): boolean => {
            if (key.startsWith("\x1b[1;") || key.startsWith("\x1b[3;")) return true;
            return false;
        },
    };

    public constructor(
        terminalName: string,
        private processCmd: (cmd: string) => Promise<string>,
        private controller: AbortController,
        options?: { startup?: string; message?: string; history?: string[]; formatCommandLine?: (cmd: string) => string }
    ) {
        this.mTerminalName = terminalName;
        this.mMessage = options?.message ?? this.mTerminalName;
        this.mHistory = options?.history ?? [];
        this.historyIndex = this.mHistory.length;
        this.command = options?.startup ?? "";
        this.charArrayCmd = [];
        this.cursorPosition = this.charArrayCmd.length;
        this.formatCommandLine = options?.formatCommandLine ?? ((cmd: string) => `${ZoweTerminal.Keys.EMPTY_LINE}${cmd}`);
        this.chalk = imperative.TextUtils.chalk;
    }

    private charArrayCmd: string[];
    private mMessage: string;
    protected mTerminalName: string = "";
    protected mHistory: string[];
    private historyIndex: number;
    private isCommandRunning = false;
    private pressedCtrlC = false;
    private chalk;

    private writeEmitter = new vscode.EventEmitter<string>();
    protected write(text: string) {
        this.writeEmitter.fire(text);
    }
    protected writeLine(text: string) {
        this.write(text);
        this.write(ZoweTerminal.Keys.NEW_LINE);
        this.writeCmd();
    }
    protected clearLine() {
        this.write(ZoweTerminal.Keys.CLEAR_LINE);
    }
    protected writeCmd(cmd?: string) {
        this.write(this.formatCommandLine ? this.formatCommandLine(cmd ?? this.command) : cmd ?? this.command);
    }
    protected refreshCmd() {
        this.command = this.sanitizeInput(this.command);
        this.pressedCtrlC = false;
        if (!this.charArrayCmd.length || this.charArrayCmd.join("") !== this.command) {
            this.charArrayCmd = Array.from(this.command);
        }
        this.clearLine();
        this.writeCmd();
        if (this.charArrayCmd.length > this.cursorPosition) {
            const getPos = (char: string) => {
                if (char === ZoweTerminal.invalidChar) return 1;
                const charBytes = Buffer.from(char).length;
                return charBytes > 2 ? 2 : 1;
            };
            const offset = this.charArrayCmd.slice(this.cursorPosition).reduce((total, curr) => total + getPos(curr), 0);
            [...Array(offset)].map(() => this.write(ZoweTerminal.Keys.LEFT));
        }
    }
    protected clear() {
        this.write(ZoweTerminal.Keys.CLEAR_ALL);
        this.writeLine(this.chalk.dim.italic(this.mMessage));
    }

    protected command: string;
    protected formatCommandLine: (cmd: string) => string;
    protected cursorPosition: number;

    public onDidWrite: vscode.Event<string> = this.writeEmitter.event;

    private closeEmitter = new vscode.EventEmitter<void>();
    public onDidClose?: vscode.Event<void> = this.closeEmitter.event;

    public open(_initialDimensions?: vscode.TerminalDimensions | undefined): void {
        this.writeLine(this.chalk.dim.italic(this.mMessage));
        if (this.command.length > 0) {
            this.handleInput(ZoweTerminal.Keys.ENTER);
        }
    }

    public close(): void {
        this.closeEmitter.fire();
    }

    private navigateHistory(offset: number): void {
        this.historyIndex = Math.max(0, Math.min(this.mHistory.length, this.historyIndex + offset));
        this.command = this.mHistory[this.historyIndex] ?? "";
        this.charArrayCmd = Array.from(this.command);
        this.cursorPosition = this.charArrayCmd.length;
        this.refreshCmd();
    }

    private moveCursor(offset: number): void {
        this.cursorPosition = Math.max(0, Math.min(this.charArrayCmd.length, this.cursorPosition + offset));
        this.refreshCmd();
    }

    private moveCursorTo(position: number): void {
        this.cursorPosition = position < 0 ? 0 : Math.min(this.charArrayCmd.length, position);
        this.refreshCmd();
    }

    private isPrintable(char: string): boolean {
        const codePoint = char.codePointAt(0);
        if (codePoint === undefined) return false;
        if (codePoint >= 0x20 && codePoint <= 0x7e) return true;
        if (codePoint >= 0xa0 && codePoint <= 0xd7ff) return true; // Control characters
        if (codePoint >= 0xe000 && codePoint <= 0xfffd) return true; // Private use area
        if (codePoint >= 0x10000 && codePoint <= 0x10ffff) return true; // Supplemental planes
        return false;
    }

    private sanitizeInput(input: string): string {
        return Array.from(input)
            .map((char) => (this.isPrintable(char) ? char : ZoweTerminal.invalidChar))
            .join("");
    }

    private deleteCharacter(offset: number): void {
        const deleteIndex = this.cursorPosition + offset;

        if (deleteIndex >= 0 && deleteIndex < this.charArrayCmd.length) {
            this.charArrayCmd.splice(deleteIndex, 1);
            this.command = this.charArrayCmd.join("");

            if (offset === -1) {
                this.cursorPosition--;
                this.write(ZoweTerminal.Keys.LEFT);
            } else if (offset === 0) {
                this.write(ZoweTerminal.Keys.DEL);
            }
            this.refreshCmd();
        }
    }

    private async handleEnter() {
        this.write(ZoweTerminal.Keys.NEW_LINE);
        const cmd = this.command;
        this.command = "";
        this.charArrayCmd = [];
        if (cmd.length === 0) {
            this.writeCmd();
            return;
        }

        if (cmd[0] === ":") {
            if (cmd === ":clear") {
                this.clear();
            } else if (cmd === ":exit") {
                this.close();
            }
        } else {
            this.isCommandRunning = true;

            const output = await Promise.race([
                this.processCmd(cmd),
                new Promise<null>((resolve, _reject) => {
                    this.controller.signal.addEventListener("abort", () => {
                        this.isCommandRunning = false;
                        resolve(null);
                    });
                    if (!this.isCommandRunning) resolve(null);
                }),
            ]);
            this.isCommandRunning = false;
            if (output === null) {
                this.writeLine(this.chalk.italic.red("Operation cancelled!"));
            } else {
                this.writeLine(output.trim().split("\n").join("\r\n"));
            }
        }
        this.mHistory.push(cmd);
        this.historyIndex = this.mHistory.length;
        this.cursorPosition = 0;
    }

    // Handle input from the terminal
    public async handleInput(data: string): Promise<void> {
        // console.log("data", data, Buffer.from(data));
        if (this.isCommandRunning) {
            if ([ZoweTerminal.Keys.CTRL_C, ZoweTerminal.Keys.CTRL_D].includes(data)) this.controller.abort();
            if (data === ZoweTerminal.Keys.CTRL_D) this.close();
            else this.pressedCtrlC = true;
            return;
        }
        if (ZoweTerminal.Keys.hasModKey(data)) return;
        switch (data) {
            case ZoweTerminal.Keys.CTRL_C:
                if (this.pressedCtrlC) this.close();
                if (this.command.length > 0) {
                    this.command = "";
                    this.handleEnter();
                } else {
                    this.writeLine(this.chalk.italic("(To exit, press Ctrl+C again or Ctrl+D or type :exit)"));
                    this.pressedCtrlC = true;
                }
                break;
            case ZoweTerminal.Keys.CTRL_D:
                this.close();
                break;
            case ZoweTerminal.Keys.UP:
            case ZoweTerminal.Keys.PAGE_UP:
                this.navigateHistory(-1);
                break;
            case ZoweTerminal.Keys.DOWN:
            case ZoweTerminal.Keys.PAGE_DOWN:
                this.navigateHistory(1);
                break;
            case ZoweTerminal.Keys.LEFT:
            case ZoweTerminal.Keys.OPT_LEFT:
                this.moveCursor(-1);
                break;
            case ZoweTerminal.Keys.RIGHT:
            case ZoweTerminal.Keys.OPT_RIGHT:
                this.moveCursor(1);
                break;
            case ZoweTerminal.Keys.HOME:
            case ZoweTerminal.Keys.CMD_LEFT:
                this.moveCursorTo(0);
                break;
            case ZoweTerminal.Keys.END:
            case ZoweTerminal.Keys.CMD_RIGHT:
                this.moveCursorTo(this.charArrayCmd.length);
                break;
            case ZoweTerminal.Keys.DEL:
                this.deleteCharacter(0);
                break;
            case ZoweTerminal.Keys.BACKSPACE:
            case ZoweTerminal.Keys.CTRL_BACKSPACE:
            case ZoweTerminal.Keys.CMD_BACKSPACE:
            case ZoweTerminal.Keys.OPT_BACKSPACE:
            case ZoweTerminal.Keys.OPT_CMD_BACKSPACE:
                this.deleteCharacter(-1);
                break;
            case ZoweTerminal.Keys.ENTER:
                await this.handleEnter();
                break;
            case ZoweTerminal.Keys.TAB:
            case ZoweTerminal.Keys.INSERT:
                // Do nothing
                break;
            default: {
                const charArray = this.charArrayCmd;
                this.command = charArray.slice(0, Math.max(0, this.cursorPosition)).join("") + data + charArray.slice(this.cursorPosition).join("");
                this.charArrayCmd = Array.from(this.command);
                this.cursorPosition = Math.min(this.charArrayCmd.length, this.cursorPosition + Array.from(data).length);
                this.write(data);
                this.refreshCmd();
            }
        }
    }
}
