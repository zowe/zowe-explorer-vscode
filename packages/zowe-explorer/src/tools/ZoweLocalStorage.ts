/**
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import * as vscode from "vscode";
import * as meta from "../../package.json";
import { ZoweLogger } from "./ZoweLogger";
import type { PersistenceSchemaEnum } from "@zowe/zowe-explorer-api";
import type { Definitions } from "../configuration/Definitions";

export class ZoweLocalStorage {
    private static storage: vscode.Memento;
    public static initializeZoweLocalStorage(state: vscode.Memento): void {
        ZoweLocalStorage.storage = state;
    }

    public static getValue<T>(key: Definitions.LocalStorageKey | PersistenceSchemaEnum): T {
        ZoweLogger.trace("ZoweLocalStorage.getValue called.");
        const defaultValue = meta.contributes.configuration.properties[key]?.default;
        return ZoweLocalStorage.storage.get<T>(key, defaultValue);
    }

    public static setValue<T>(key: Definitions.LocalStorageKey | PersistenceSchemaEnum, value: T): Thenable<void> {
        ZoweLogger.trace("ZoweLocalStorage.setValue called.");
        return ZoweLocalStorage.storage.update(key, value);
    }
}
