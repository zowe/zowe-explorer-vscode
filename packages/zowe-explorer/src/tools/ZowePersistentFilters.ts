/**
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { PersistenceSchemaEnum } from "@zowe/zowe-explorer-api";
import { Constants } from "../configuration/Constants";
import { ZoweLogger } from "./ZoweLogger";
import { ZoweLocalStorage } from "./ZoweLocalStorage";
import { Definitions } from "../configuration/Definitions";

/**
 * Standard history and favorite persistance handling routines
 *
 * @export
 * @class PersistentFilters
 */
export class ZowePersistentFilters {
    private static readonly favorites: string = "favorites";
    private static readonly searchHistory: string = "searchHistory";
    private static readonly fileHistory: string = "fileHistory";
    private static readonly sessions: string = "sessions";
    private static readonly encodingHistory: string = "encodingHistory";

    public schema: PersistenceSchemaEnum;
    private mSearchHistory: string[] = [];
    private mFileHistory: string[] = [];
    private mSessions: string[] = [];
    private mEncodingHistory: string[] = [];

    public constructor(
        schema: PersistenceSchemaEnum,
        private maxSearchHistory = Constants.MAX_SEARCH_HISTORY,
        private maxFileHistory = Constants.MAX_FILE_HISTORY
    ) {
        ZoweLogger.trace("PersistentFilters.constructor called.");
        this.schema = schema;
        this.initialize();
    }

    /*********************************************************************************************************************************************/
    /* Add functions, for adding items to the persistent settings
    /*********************************************************************************************************************************************/

    /**
     * Adds one line of search history to the local store and
     * updates persistent store. The store contains a
     * maximum number of entries as described by `maxSearchHistory`
     *
     * If the entry matches a previous entry it is removed from the list
     * at that position in the stack.
     *
     * Once the maximum capacity has been reached the last entry is popped off
     *
     * @param {string} criteria - a line of search criteria
     */
    public addSearchHistory(criteria: string): void {
        ZoweLogger.trace("PersistentFilters.addSearchHistory called.");
        if (criteria) {
            // Remove any entries that match
            this.mSearchHistory = this.mSearchHistory.filter((element) => {
                return element.trim() !== criteria.trim();
            });

            // Add value to front of stack
            this.mSearchHistory.unshift(criteria);

            // If list getting too large remove last entry
            if (this.mSearchHistory.length > this.maxSearchHistory) {
                this.mSearchHistory.pop();
            }
            this.updateSearchHistory();
        }
    }

    /**
     * Adds the name of one recently-edited file to the local store and
     * updates persistent store. The store contains a
     * maximum number of entries as described by `maxFileHistory`
     *
     * If the entry matches a previous entry it is removed from the list
     * at that position in the stack.
     *
     * Once the maximum capacity has been reached the last entry is popped off
     *
     * @param {string} criteria - a line of search criteria
     */
    public addFileHistory(criteria: string): void {
        ZoweLogger.trace("PersistentFilters.addFileHistory called.");
        if (criteria) {
            criteria = criteria.toUpperCase();
            // Remove any entries that match
            this.mFileHistory = this.mFileHistory.filter((element) => {
                return element.trim() !== criteria.trim();
            });

            // Add value to front of stack
            this.mFileHistory.unshift(criteria);

            // If list getting too large remove last entry
            if (this.mFileHistory.length > this.maxFileHistory) {
                this.mFileHistory.pop();
            }
            this.updateFileHistory();
        }
    }

    /**
     * Adds one line of session history to the local store and
     * updates persistent store.
     *
     * If the entry matches a previous entry it is removed from the list
     * at that position in the stack.
     *
     * @param {string} criteria - a session name
     */
    public addSession(criteria: string): void {
        ZoweLogger.trace("PersistentFilters.addSession called.");
        // Remove any entries that match
        this.mSessions = this.mSessions.filter((element) => {
            return element.trim() !== criteria.trim();
        });
        this.mSessions.push(criteria);

        // Use standard sorting
        this.mSessions.sort((a, b) => a.localeCompare(b));
        this.updateSessions();
    }

    /*********************************************************************************************************************************************/
    /* Get/read functions, for returning the values stored in the persistent arrays
    /*********************************************************************************************************************************************/

    public getSearchHistory(): string[] {
        ZoweLogger.trace("PersistentFilters.getSearchHistory called.");
        return this.mSearchHistory;
    }

    public getSessions(): string[] {
        ZoweLogger.trace("PersistentFilters.getSessions called.");
        return this.mSessions;
    }

    public getFileHistory(): string[] {
        ZoweLogger.trace("PersistentFilters.getFileHistory called.");
        return this.mFileHistory;
    }

    public readFavorites(): string[] {
        ZoweLogger.trace("PersistentFilters.readFavorites called.");
        const localStorageSchema = ZoweLocalStorage.getValue<Definitions.ZowePersistentFilter>(this.schema);
        if (localStorageSchema) {
            return localStorageSchema[ZowePersistentFilters.favorites] as string[];
        }
        return [];
    }

    /*********************************************************************************************************************************************/
    /* Remove functions, for removing one item from the persistent arrays
    /*********************************************************************************************************************************************/

    public removeSession(name: string): void {
        ZoweLogger.trace("PersistentFilters.removeSession called.");
        // Remove any entries that match
        this.mSessions = this.mSessions.filter((element) => {
            return element.trim() !== name.trim();
        });
        this.updateSessions();
    }

    /**
     * @param name - Should be in format "[session]: DATASET.QUALIFIERS" or "[session]: /file/path", as appropriate
     */
    public removeFileHistory(name: string): void {
        const index = this.mFileHistory.findIndex((fileHistoryItem) => {
            return fileHistoryItem.includes(name.toUpperCase());
        });
        if (index >= 0) {
            this.mFileHistory.splice(index, 1);
        }
        return this.updateFileHistory();
    }

    public removeSearchHistory(name: string): void {
        const index = this.mSearchHistory.findIndex((searchHistoryItem) => {
            return searchHistoryItem.includes(name);
        });
        if (index >= 0) {
            this.mSearchHistory.splice(index, 1);
        }
        return this.updateSearchHistory();
    }

    public removeEncodingHistory(name: string): void {
        const index = this.mEncodingHistory.findIndex((encodingHistoryItem) => {
            return encodingHistoryItem.includes(name);
        });
        if (index >= 0) {
            this.mEncodingHistory.splice(index, 1);
        }
        this.updateEncodingHistory();
    }

    /*********************************************************************************************************************************************/
    /* Reset functions, for resetting the persistent array to empty (in the extension and in settings.json)
    /*********************************************************************************************************************************************/

    public resetSearchHistory(): void {
        ZoweLogger.trace("PersistentFilters.resetSearchHistory called.");
        this.mSearchHistory = [];
        this.updateSearchHistory();
    }

    public resetSessions(): void {
        ZoweLogger.trace("PersistentFilters.resetSessions called.");
        this.mSessions = [];
        this.updateSessions();
    }

    public resetFileHistory(): void {
        ZoweLogger.trace("PersistentFilters.resetFileHistory called.");
        this.mFileHistory = [];
        this.updateFileHistory();
    }

    public resetEncodingHistory(): void {
        this.mEncodingHistory = [];
        this.updateEncodingHistory();
    }

    /*********************************************************************************************************************************************/
    /* Update functions, for updating the settings.json file in VSCode
    /*********************************************************************************************************************************************/

    public updateFavorites(favorites: string[]): void {
        ZoweLogger.trace("PersistentFilters.updateFavorites called.");
        const settings = ZoweLocalStorage.getValue<Definitions.ZowePersistentFilter>(this.schema);
        if (settings.persistence) {
            settings.favorites = favorites;
            ZoweLocalStorage.setValue<Definitions.ZowePersistentFilter>(this.schema, settings);
        }
    }

    private updateSearchHistory(): void {
        ZoweLogger.trace("PersistentFilters.updateSearchHistory called.");
        const settings = { ...ZoweLocalStorage.getValue<Definitions.ZowePersistentFilter>(this.schema) };
        if (settings.persistence) {
            settings.searchHistory = this.mSearchHistory;
            ZoweLocalStorage.setValue<Definitions.ZowePersistentFilter>(this.schema, settings);
        }
    }

    private updateSessions(): void {
        ZoweLogger.trace("PersistentFilters.updateSessions called.");
        const settings = { ...ZoweLocalStorage.getValue<Definitions.ZowePersistentFilter>(this.schema) };
        if (settings.persistence) {
            settings.sessions = this.mSessions;
            ZoweLocalStorage.setValue<Definitions.ZowePersistentFilter>(this.schema, settings);
        }
    }

    private updateFileHistory(): void {
        ZoweLogger.trace("PersistentFilters.updateFileHistory called.");
        const settings = { ...ZoweLocalStorage.getValue<Definitions.ZowePersistentFilter>(this.schema) };
        if (settings.persistence) {
            settings.fileHistory = this.mFileHistory;
            ZoweLocalStorage.setValue<Definitions.ZowePersistentFilter>(this.schema, settings);
        }
    }

    private updateEncodingHistory(): void {
        const settings = { ...ZoweLocalStorage.getValue<Definitions.ZowePersistentFilter>(this.schema) };
        if (settings.persistence) {
            settings.encodingHistory = this.mEncodingHistory;
            ZoweLocalStorage.setValue<Definitions.ZowePersistentFilter>(this.schema, settings);
        }
    }

    private initialize(): void {
        ZoweLogger.trace("PersistentFilters.initialize called.");
        const settings = ZoweLocalStorage.getValue<Definitions.ZowePersistentFilter>(this.schema);
        if (settings) {
            this.mSearchHistory = settings[ZowePersistentFilters.searchHistory] ?? [];
            this.mSessions = settings[ZowePersistentFilters.sessions] ?? [];
            this.mFileHistory = settings[ZowePersistentFilters.fileHistory] ?? [];
            this.mEncodingHistory = settings[ZowePersistentFilters.encodingHistory] ?? [];
        }
        this.updateSearchHistory();
        this.updateSessions();
        this.updateFileHistory();
        this.updateEncodingHistory();
    }
}
