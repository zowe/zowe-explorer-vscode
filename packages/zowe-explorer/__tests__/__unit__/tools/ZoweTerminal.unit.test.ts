/**
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { ZoweTerminal } from "../../../src/tools/ZoweTerminal";

describe("ZoweTerminal Unit Tests", () => {
    it("should not change the keys", () => {
        expect(ZoweTerminal.Keys).toMatchSnapshot();
    });

    it("should handle ctrl_c to cancel a running command", async () => {
        const spyCb = jest.fn().mockImplementation(async (cmd: string) => Promise.resolve("test-output"));
        const signalSpy = jest.fn().mockImplementation((_event, cb) => cb());
        const iTerm = new ZoweTerminal("test", spyCb, { abort: jest.fn(), signal: { addEventListener: signalSpy } } as any, { history: ["old"] });
        (iTerm as any).command = "open";
        iTerm.open();

        await iTerm.handleInput(ZoweTerminal.Keys.ENTER);
        (iTerm as any).isCommandRunning = true;
        await iTerm.handleInput(ZoweTerminal.Keys.CTRL_D);
        expect(spyCb).toHaveBeenCalledWith("open");
        spyCb.mockClear();

        expect((iTerm as any).mHistory as string[]).toEqual(["old", "open"]);
    });

    it("should send the entered command to the callback function", async () => {
        const spyCb = jest.fn().mockImplementation(async (cmd: string) => Promise.resolve("test-output"));
        const iTerm = new ZoweTerminal("test", spyCb, { signal: { addEventListener: jest.fn() } } as any, { history: ["old"] });
        iTerm.open();

        await iTerm.handleInput("testABC");
        await iTerm.handleInput(ZoweTerminal.Keys.ENTER);
        expect(spyCb).toHaveBeenCalledWith("testABC");
        spyCb.mockClear();

        await iTerm.handleInput(ZoweTerminal.Keys.HOME); // |testABC
        await iTerm.handleInput(ZoweTerminal.Keys.END); // testABC|
        await iTerm.handleInput(ZoweTerminal.Keys.CMD_LEFT); // |testABC
        await iTerm.handleInput(ZoweTerminal.Keys.CMD_RIGHT); // testABC|
        await iTerm.handleInput(ZoweTerminal.Keys.UP); // testABC|
        await iTerm.handleInput(ZoweTerminal.Keys.UP); // old|
        await iTerm.handleInput(ZoweTerminal.Keys.DOWN); // testABC|
        await iTerm.handleInput(ZoweTerminal.Keys.LEFT); // testAB|C
        await iTerm.handleInput(ZoweTerminal.Keys.LEFT); // testA|BC
        await iTerm.handleInput(ZoweTerminal.Keys.BACKSPACE); // test|BC
        await iTerm.handleInput(ZoweTerminal.Keys.RIGHT); // testB|C
        await iTerm.handleInput(ZoweTerminal.Keys.BACKSPACE); // test|C
        // handle multiple characters in sequence (CPU delay / copy+paste)
        await iTerm.handleInput("1A"); // test1A|C
        await iTerm.handleInput(ZoweTerminal.Keys.BACKSPACE); // test1|C
        // Handle double byte characters
        await iTerm.handleInput("🙏🙏"); // test1🙏🙏|C
        await iTerm.handleInput(ZoweTerminal.Keys.BACKSPACE); // test1🙏|C
        // Handle unicode "Hello"
        await iTerm.handleInput("\u0048\u0065\u006C\u006C\u006F"); // test1🙏Hello|C
        await iTerm.handleInput(ZoweTerminal.Keys.DEL); // test1🙏Hello|
        await iTerm.handleInput(ZoweTerminal.Keys.ENTER);
        expect(spyCb).toHaveBeenCalledWith("test1🙏Hello");
        spyCb.mockClear();

        (iTerm as any).command = "";
        await iTerm.handleInput(ZoweTerminal.Keys.INSERT); // do nothing
        await iTerm.handleInput(ZoweTerminal.Keys.TAB); // do nothing
        await iTerm.handleInput("\x1b[1;A"); // Shift+Up // do nothing
        await iTerm.handleInput("\x1b[3;A"); // fn+option+shift+up // do nothing
        await iTerm.handleInput(ZoweTerminal.Keys.ENTER);
        await iTerm.handleInput(ZoweTerminal.Keys.UP); // "test1🙏Hello"
        await iTerm.handleInput(ZoweTerminal.Keys.CTRL_C); // Clear the terminal
        await iTerm.handleInput(ZoweTerminal.Keys.CTRL_C); // Present the "(to exit ...)" message
        await iTerm.handleInput(ZoweTerminal.Keys.CTRL_C); // close
        await iTerm.handleInput(ZoweTerminal.Keys.CTRL_D); // close
        await iTerm.handleInput(":clear");
        await iTerm.handleInput(ZoweTerminal.Keys.ENTER);
        await iTerm.handleInput(":exit");
        await iTerm.handleInput(ZoweTerminal.Keys.ENTER);

        expect((iTerm as any).mHistory as string[]).toEqual(["old", "testABC", "test1🙏Hello", ":clear", ":exit"]);
    });
});
