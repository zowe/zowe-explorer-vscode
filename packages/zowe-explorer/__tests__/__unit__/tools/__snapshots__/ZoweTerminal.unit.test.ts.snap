// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZoweTerminal Unit Tests should not change the keys 1`] = `
{
  "BACKSPACE": "",
  "CLEAR_ALL": "[2J[3J[;H",
  "CLEAR_LINE": "[2K
",
  "CMD_BACKSPACE": "",
  "CMD_LEFT": "",
  "CMD_RIGHT": "",
  "CTRL_BACKSPACE": "",
  "CTRL_C": "",
  "CTRL_D": "",
  "DEL": "[3~",
  "DOWN": "[B",
  "EMPTY_LINE": "> ",
  "END": "[F",
  "ENTER": "
",
  "HOME": "[H",
  "INSERT": "[2~",
  "LEFT": "[D",
  "NEW_LINE": "
",
  "OPT_BACKSPACE": "",
  "OPT_CMD_BACKSPACE": "",
  "OPT_LEFT": "b",
  "OPT_RIGHT": "f",
  "PAGE_DOWN": "[6~",
  "PAGE_UP": "[5~",
  "RIGHT": "[C",
  "TAB": "	",
  "UP": "[A",
  "hasModKey": [Function],
}
`;
